moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,375 @@
+            return invoke(intermediate)
+          }
+        }
+      } else {
+        throw IllegalArgumentException(
+          """Unexpected signature for $method.
+@FromJson method signatures may have one of the following structures:
+    <any access modifier> R fromJson(JsonReader jsonReader) throws <any>;
+    <any access modifier> R fromJson(JsonReader jsonReader, JsonAdapter<any> delegate, <any more delegates>) throws <any>;
+    <any access modifier> R fromJson(T value) throws <any>;
+"""
+        )
+      }
+    }
+
+    /** Returns the matching adapter method from the list.  */
+    private fun get(
+      adapterMethods: List<AdapterMethod>,
+      type: Type,
+      annotations: Set<Annotation>
+    ): AdapterMethod? {
+      var i = 0
+      val size = adapterMethods.size
+      while (i < size) {
+        val adapterMethod = adapterMethods[i]
+        if (Types.equals(adapterMethod.type, type) && adapterMethod.annotations == annotations) {
+          return adapterMethod
+        }
+        i++
+      }
+      return null
+    }
+  }
+
+  internal abstract class AdapterMethod(
+    type: Type,
+    val annotations: Set<Annotation>,
+    val adapter: Any,
+    val method: Method,
+    parameterCount: Int,
+    adaptersOffset: Int,
+    val nullable: Boolean
+  ) {
+    val type: Type
+    private val adaptersOffset: Int
+    private val jsonAdapters: Array<JsonAdapter<*>?>
+
+    init {
+      this.type = canonicalize(type)
+      this.adaptersOffset = adaptersOffset
+      jsonAdapters = arrayOfNulls(parameterCount - adaptersOffset)
+    }
+
+    open fun bind(moshi: Moshi, factory: JsonAdapter.Factory) {
+      if (jsonAdapters.isNotEmpty()) {
+        val parameterTypes = method.genericParameterTypes
+        val parameterAnnotations = method.parameterAnnotations
+        var i = adaptersOffset
+        val size = parameterTypes.size
+        while (i < size) {
+          val type = (parameterTypes[i] as ParameterizedType).actualTypeArguments[0]
+          val jsonAnnotations = jsonAnnotations(parameterAnnotations[i])
+          jsonAdapters[i - adaptersOffset] =
+            if (Types.equals(this.type, type) && annotations == jsonAnnotations) {
+              moshi.nextAdapter(
+                factory, type, jsonAnnotations
+              )
+            } else {
+              moshi.adapter<Any>(type, jsonAnnotations)
+            }
+          i++
+        }
+      }
+    }
+
+    open fun toJson(moshi: Moshi, writer: JsonWriter, value: Any?) {
+      throw AssertionError()
+    }
+
+    open fun fromJson(moshi: Moshi, reader: JsonReader): Any? {
+      throw AssertionError()
+    }
+
+    /** Invoke the method with one fixed argument, plus any number of JSON adapter arguments.  */
+    protected fun invoke(a1: Any?): Any? {
+      val args = arrayOfNulls<Any>(1 + jsonAdapters.size)
+      args[0] = a1
+      System.arraycopy(jsonAdapters, 0, args, 1, jsonAdapters.size)
```

Comment:
There's a kotlin alternative to this right?

Reply:
https://github.com/square/moshi/pull/1467/commits/002ae5911334aa454215e8a92d9d2650f6030224#diff-e8a15c8ca843b45b53cf5d1b2097710c70dd4b5a9399c074e7fa2b379a772d47R339